### PR TITLE
Feature/porter runner migrate to server refactor

### PR DESCRIPTION
**Pull request by TedxTed**
migrate backend porter runner to server 

### Porting Last Logged In Data:
 - Added functionality to retrieve and update the last login timestamps of users from cached data.
 ### Porting Player Event Data:
 - retrieve player event data from the cache and store it in the database
 ### Porting Podcast Program Data
- extracting podcast program progress data from the cache and storing it in the database.

## others
- Batch Processing Logic

### CODE REVIEW

**Comment by Ponchimeow** [comment]
[(2)【歷程記錄】會員觀看課程我的主頁進度條與歷程不同步](https://trello.com/c/m2cECZmJ/4820-2%E3%80%90%E6%AD%B7%E7%A8%8B%E8%A8%98%E9%8C%84%E3%80%91%E6%9C%83%E5%93%A1%E8%A7%80%E7%9C%8B%E8%AA%B2%E7%A8%8B%E6%88%91%E7%9A%84%E4%B8%BB%E9%A0%81%E9%80%B2%E5%BA%A6%E6%A2%9D%E8%88%87%E6%AD%B7%E7%A8%8B%E4%B8%8D%E5%90%8C%E6%AD%A5)

File: src/runner/porter.runner.ts [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        const [, memberId, , programContentId] = key.split(':');
+        const value = JSON.parse(valueString || '{}');
+
+        try {
+          const programContentRepo = manager.getRepository(ProgramContent);
+          const programContent = await programContentRepo.findOneBy({ id: programContentId });
+          const programContentLog = new ProgramContentLog();
+          programContentLog.memberId = memberId;
+          programContentLog.programContent = programContent;
+          programContentLog.playbackRate = value.playbackRate;
```
Comment (aniku777tw) [inline src/runner/porter.runner.ts]: if playback rate not exist, please default this value by 1

File: src/runner/porter.runner.ts [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    const programContentLogs: ProgramContentLog[] = [];
+
+    for (let batchIndex = 0; batchIndex < allKeys.length; batchIndex += batchSize) {
+      const batchKeys = allKeys.slice(batchIndex, batchIndex + batchSize);
+      const values = await this.cacheService.getClient().mget(batchKeys);
+
+      for (let keyIndex = 0; keyIndex < batchKeys.length; keyIndex++) {
+        const key = batchKeys[keyIndex];
+        const valueString = values[keyIndex];
+        const [, memberId, , programContentId] = key.split(':');
```
Comment (aniku777tw) [inline src/runner/porter.runner.ts]: we have the key created time (unix) in cache key, use
```
const [, memberId, , podcastProgramId, createdAt ] = key.split(':');
```
and use this ```createdAt``` to insert program content log

File: test/runner/porter_runner.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        podcastAlbumId: `${podcastAlbum.id}`,
+      }),
+      'EX',
+      7 * 86400,
+    );
+
+    await application.init();
+  });
+
+  afterAll(async () => {
```
Comment (aniku777tw) [inline test/runner/porter_runner.e2e-spec.ts]: delete all data before close app
```
    await programPlanRepo.delete({});
    await currencyRepo.delete({});
    await programContentLogRepo.delete({});
    await programContentProgressRepo.delete({});
    await programContentRepo.delete({});
    await programContentSectionRepo.delete({});
    await programContentBodyRepo.delete({});
    await programRepo.delete({});
    await podcastProgramProgressRepo.delete({});
    await podcastProgramRepo.delete({});
    await podcastAlbumRepo.delete({});
    await memberRepo.delete({});
    await appSettingRepo.delete({});
    await appSecretRepo.delete({});
    await appHostRepo.delete({});
    await appRepo.delete({});
    await appPlanRepo.delete({});
    await roleRepo.delete({});
```

File: src/runner/porter.runner.ts [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+      let loginedAt: string | null = memberLastLoggedInTimestamps[index];
+
+      if (loginedAt) {
+        const localDate = new Date(loginedAt);
+        loginedAt = localDate.toISOString();
+      }
+
+      try {
+        await manager.update(Member, memberId, { loginedAt });
```
Comment (aniku777tw) [inline src/runner/porter.runner.ts]: In my opinion to create a method in MemberInfra.ts  is better

File: src/runner/porter.runner.ts [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+        await manager.findOne(Member, { where: { id: memberId } });
```
Comment (aniku777tw) [inline src/runner/porter.runner.ts]: unnecessary query

File: src/runner/porter.runner.ts [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+      for (let keyIndex = 0; keyIndex < batchKeys.length; keyIndex++) {
+        const key = batchKeys[keyIndex];
+        const valueString = values[keyIndex];
+        const [, memberId, , programContentId] = key.split(':');
+        const value = JSON.parse(valueString || '{}');
+
+        try {
+          const programContentRepo = manager.getRepository(ProgramContent);
+          const programContent = await programContentRepo.findOneBy({ id: programContentId });
```
Comment (aniku777tw) [inline src/runner/porter.runner.ts]: create a ```programContentLogInfra.ts``` to handle will be more reusable

File: src/runner/porter.runner.ts [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    try {
+      for (const [, info] of progressMap) {
+        let podcastProgramProgress = await manager.getRepository(PodcastProgramProgress).findOne({
+          where: {
+            memberId: info.memberId,
+            podcastProgramId: info.podcastProgramId,
+          },
+        });
```
Comment (aniku777tw) [inline src/runner/porter.runner.ts]: same thing as above

File: src/runner/porter.runner.ts [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+        if (!podcastProgramProgress) {
+          podcastProgramProgress = manager.getRepository(PodcastProgramProgress).create(info);
```
Comment (aniku777tw) [inline src/runner/porter.runner.ts]: same thing as above

File: test/runner/porter_runner.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    await podcastAlbumRepo.delete({});
+    await currencyRepo.delete({});
+    await memberRepo.delete({});
+    await appHostRepo.delete({});
+    await appSecretRepo.delete({});
+    await appSettingRepo.delete({});
+    await appRepo.delete({});
+    await appPlanRepo.delete({});
+    await roleRepo.delete({});
+
```
Comment (aniku777tw) [inline test/runner/porter_runner.e2e-spec.ts:192]: please clear redis cache before close app

**Review by aniku777tw — CHANGES_REQUESTED** [review]

File: src/runner/porter.runner.ts (hunk 6 above)
Comment (TedxTed) [reply src/runner/porter.runner.ts]: Sure, I'll use the existing program.infra.ts to implement it first, if you think it is not proper, I'll make programcontent.infra independently.

File: src/runner/porter.runner.ts (hunk 5 above)
Comment (TedxTed) [reply src/runner/porter.runner.ts]: yes, you are right , the query to find the member after updating it is unnecessary. I'll remove it , thanks

File: src/runner/porter.runner.ts (hunk 4 above)
Comment (TedxTed) [reply src/runner/porter.runner.ts]: sure , i will move crud to infra ～

File: test/runner/porter_runner.e2e-spec.ts (hunk 9 above)
Comment (TedxTed) [reply test/runner/porter_runner.e2e-spec.ts:192]: ok , i add `await cacheService.getClient().flushall();` to `afterEach` hook

File: test/runner/porter_runner.e2e-spec.ts (hunk 3 above)
Comment (TedxTed) [reply test/runner/porter_runner.e2e-spec.ts]: ok ,  I'll include this in the `afterEach` hook

File: src/runner/porter.runner.ts (hunk 2 above)
Comment (TedxTed) [reply src/runner/porter.runner.ts]: ok , I'll used the created time from the cache key~

**Comment by clothe09986** [comment]
[【DC alert】將porter runner移到 lodestar-server，解決alert一直響](https://trello.com/c/1bPejRQh/4871-%E3%80%90dc-alert%E3%80%91%E5%B0%87porter-runner%E7%A7%BB%E5%88%B0-lodestar-server%EF%BC%8C%E8%A7%A3%E6%B1%BAalert%E4%B8%80%E7%9B%B4%E9%9F%BF)

**Review by aniku777tw — APPROVED** [review]

File: src/runner/porter.runner.ts (hunk 2 above)
Comment (kkshyu) [reply src/runner/porter.runner.ts]: add a comment like 
```
// key format:  `program-content:{memberId}:XXXX:{podcastProgramId}:createdAt`
```

**Pull request merged**